### PR TITLE
Performance: Avoid Repeat Calls to getCredentials()

### PR DESCRIPTION
**Pull request by iambmelt**
This PR aims to improve the performance of `getAccounts()`.

By avoiding repeated calls to `getCredentials()` we can avoid iterating over the cache + keys multiple times.

>Note: this logic will still result in multiple `List` traversals of the cached contents when calling `getCredentialsFilteredByInternal()` however, we avoid any repeat decryption/cache accesses with this change. Refactoring  `getCredentialsFilteredByInternal()` to directly accept a `Set<CredentialType>` is a less straightforward change, so not attempting that as part of this PR.

Cache tests continue to pass with this change.

### CODE REVIEW

File: common/src/main/java/com/microsoft/identity/common/internal/cache/IAccountCredentialCache.java [hunk 1]
```diff
@@ -135,6 +136,17 @@
             final String requestedClaims
     );
 
+    List<Credential> getCredentialsFilteredBy(
```
Comment (iambmelt) [inline common/src/main/java/com/microsoft/identity/common/internal/cache/IAccountCredentialCache.java:139]: javadoc

File: common/src/main/java/com/microsoft/identity/common/internal/cache/SharedPreferencesAccountCredentialCache.java [hunk 2]
```diff
@@ -375,6 +375,38 @@ public boolean test(String value) {
...
+    public List<Credential> getCredentialsFilteredBy(@Nullable final String homeAccountId,
+                                                     @Nullable final String environment,
+                                                     @NonNull final Set<CredentialType> credentialTypes,
+                                                     @Nullable final String clientId,
+                                                     @Nullable final String realm,
+                                                     @Nullable final String target,
+                                                     @Nullable final String authScheme,
+                                                     @Nullable final String requestedClaims) {
+        final List<Credential> result = new ArrayList<>();
+        final List<Credential> allCredentials = getCredentials();
```
Comment (iambmelt) [inline common/src/main/java/com/microsoft/identity/common/internal/cache/SharedPreferencesAccountCredentialCache.java:388]: Flip declaration order

Comment (iambmelt) [reply common/src/main/java/com/microsoft/identity/common/internal/cache/SharedPreferencesAccountCredentialCache.java:381]: b5db566de1c134176ecbedca930b811dad89d749

File: common/src/main/java/com/microsoft/identity/common/internal/cache/IAccountCredentialCache.java (hunk 1 above)
Comment (iambmelt) [reply common/src/main/java/com/microsoft/identity/common/internal/cache/IAccountCredentialCache.java:151]: ac900823f3cc80baa52af8acc76c606a1c732e39

File: common/src/main/java/com/microsoft/identity/common/internal/cache/SharedPreferencesAccountCredentialCache.java [hunk 3]
```diff
@@ -283,20 +280,17 @@ public boolean test(String value) {
...
-
-                if (null == credential) {
-                    Logger.warn(
-                            TAG,
-                            CREDENTIAL_DESERIALIZATION_FAILED
-                    );
-                } else {
-                    credentials.put(cacheKey, credential);
-                }
+            final Credential credential = mCacheValueDelegate.fromCacheValue(
```
Comment (iambmelt) [inline common/src/main/java/com/microsoft/identity/common/internal/cache/SharedPreferencesAccountCredentialCache.java:283]: This is just an indentation fix.

**Review by shahzaibj — APPROVED** [review]
LGTM!

**Review by AdamBJohnsonx — APPROVED** [review]

**Pull request merged**